pkg/annotations/middleware.go

```diff
@@ -3,65 +3,102 @@ package annotations
-	annotationRateLimitBurst           = baseAnnotation + "ratelimit-burst"
-)
+type middlewareBuilder func(middleware *dynamic.Middleware, annotations map[string]string) (bool, error)
 
 // BuildMiddleware builds a middleware from the given annotations
 func BuildMiddleware(annotations map[string]string) (*dynamic.Middleware, error) {
-	var middleware dynamic.Middleware
-
-	// Build circuit-breaker middleware.
-	if circuitBreakerExpression, ok := annotations[annotationCircuitBreakerExpression]; ok {
-		middleware.CircuitBreaker = &dynamic.CircuitBreaker{Expression: circuitBreakerExpression}
+	builders := []middlewareBuilder{
+		buildRetryMiddleware,
+		buildRateLimitMiddleware,
+		buildCircuitBreakerMiddleware,
 	}
 
-	// Build retry middleware.
-	if retryAttempts, ok := annotations[annotationRetryAttempts]; ok {
-		attempts, err := strconv.Atoi(retryAttempts)
+	var (
+		middleware        dynamic.Middleware
+		middlewareCounter int
+	)
+
+	for _, builder := range builders {
+		ok, err := builder(&middleware, annotations)
 		if err != nil {
-			return nil, fmt.Errorf("unable to build retry middleware, %q annotation is invalid: %w", annotationRetryAttempts, err)
+			return nil, err
 		}
 
-		middleware.Retry = &dynamic.Retry{Attempts: attempts}
+		if ok {
+			middlewareCounter++
+		}
 	}
 
-	// Build rate-limit middleware.
-	rateLimitAverage, hasRateLimitAverage := annotations[annotationRateLimitAverage]
-	rateLimitBurst, hasRateLimitBurst := annotations[annotationRateLimitBurst]
+	if middlewareCounter == 0 {
+		return nil, nil
+	}
 
-	if hasRateLimitAverage && hasRateLimitBurst {
-		average, err := strconv.Atoi(rateLimitAverage)
-		if err != nil {
-			return nil, fmt.Errorf("unable to build rate-limit middleware, %q annotation is invalid: %w", annotationRateLimitAverage, err)
-		}
+	return &middleware, nil
+}
 
-		burst, err := strconv.Atoi(rateLimitBurst)
-		if err != nil {
-			return nil, fmt.Errorf("unable to build rate-limit middleware, %q annotation is invalid: %w", annotationRateLimitBurst, err)
+func buildRetryMiddleware(middleware *dynamic.Middleware, annotations map[string]string) (bool, error) {
+	retryAttempts, errRetryAttempts := GetRetryAttempts(annotations)
+	if errRetryAttempts != nil {
+		if errRetryAttempts == ErrNotFound {
+			return false, nil
 		}
 
-		if burst <= 0 || average <= 0 {
-			return nil, errors.New("unable to build rate-limit middleware, burst and average must be greater than 0")
-		}
+		return false, fmt.Errorf("unable to build retry middleware: %w", errRetryAttempts)
+	}
 
-		middleware.RateLimit = &dynamic.RateLimit{
-			Average: int64(average),
-			Burst:   int64(burst),
-		}
+	middleware.Retry = &dynamic.Retry{Attempts: retryAttempts}
+
+	return true, nil
+}
+
+func buildRateLimitMiddleware(middleware *dynamic.Middleware, annotations map[string]string) (bool, error) {
+	var (
+		rateLimitBurst   int
+		rateLimitAverage int
+		err              error
+	)
+
+	rateLimitBurst, err = GetRateLimitBurst(annotations)
```

Comment:
There's no shadow declaration risk anymore since we're in a separate func.

```suggestion
	rateLimitBurst, err := GetRateLimitBurst(annotations)
```